File: studypulse/mobile/lib/models/exam.dart
class Exam {
  final int id;
  final String name;
  final String? description;
  final String? icon;
  final List<Subject> subjects;

  Exam({
    required this.id,
    required this.name,
    this.description,
    this.icon,
    this.subjects = const [],
  });

  factory Exam.fromJson(Map<String, dynamic> json) {
    return Exam(
      id: json['id'] as int,
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      icon: json['icon'] as String?,
      subjects: (json['subjects'] as List<dynamic>?)
              ?.map((s) => Subject.fromJson(s as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class Subject {
  final int id;
  final String name;
  final String? description;
  final int examId;
  final List<Topic> topics;

  Subject({
    required this.id,
    required this.name,
    this.description,
    required this.examId,
    this.topics = const [],
  });

  factory Subject.fromJson(Map<String, dynamic> json) {
    return Subject(
      id: json['id'] as int,
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      examId: json['exam_id'] as int? ?? 0,
      topics: (json['topics'] as List<dynamic>?)
              ?.map((t) => Topic.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class Topic {
  final int id;
  final String name;
  final String? description;
  final int subjectId;
  final int? questionCount;

  Topic({
    required this.id,
    required this.name,
    this.description,
    required this.subjectId,
    this.questionCount,
  });

  factory Topic.fromJson(Map<String, dynamic> json) {
    return Topic(
      id: json['id'] as int,
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      subjectId: json['subject_id'] as int? ?? 0,
      questionCount: json['question_count'] as int?,
    );
  }
}

File: studypulse/backend/app/schemas/exam.py
"""Pydantic schemas for exam, subject, and topic operations."""
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import datetime


class TopicResponse(BaseModel):
    """Schema for topic response data."""
    id: int
    subject_id: int
    name: str
    description: Optional[str] = None
    difficulty_level: str = "medium"
    estimated_study_mins: int = 60
    icon_url: Optional[str] = None
    question_count: int = 0
    
    model_config = ConfigDict(from_attributes=True)


class SubjectResponse(BaseModel):
    """Schema for subject response data."""
    id: int
    exam_id: int
    name: str
    description: Optional[str] = None
    syllabus_url: Optional[str] = None
    icon_url: Optional[str] = None
    topics: List[TopicResponse] = []
    
    model_config = ConfigDict(from_attributes=True)


class SubjectBrief(BaseModel):
    """Brief subject info without topics."""
    id: int
    exam_id: int
    name: str
    icon_url: Optional[str] = None
    topic_count: int = 0
    
    model_config = ConfigDict(from_attributes=True)


class ExamResponse(BaseModel):
    """Schema for exam response data."""
    id: int
    name: str
    description: Optional[str] = None
    category: Optional[str] = None
    conducting_body: Optional[str] = None
    exam_duration_mins: int = 180
    total_questions: int = 100
    icon_url: Optional[str] = None
    subjects: List[SubjectBrief] = []
    
    model_config = ConfigDict(from_attributes=True)


class ExamBrief(BaseModel):
    """Brief exam info for listing."""
    id: int
    name: str
    category: Optional[str] = None
    icon_url: Optional[str] = None
    subject_count: int = 0
    
    model_config = ConfigDict(from_attributes=True)
